### ISSUE TITLE
Support more granular test decoration

### ISSUE THREAD

**Issue by jbschlosser** [pytorch/pytorch#79161]
Currently, the `DecorateInfo` class only supports conditionally applying decorators to test functions based on test class name, base test name, device, and dtype. Example usage:

https://github.com/pytorch/pytorch/blob/7cb4a76844edfed7e31c50ec5af8fbfe8cbb6ca0/torch/testing/_internal/common_methods_invocations.py#L9800-L9805

It would be useful to support more granular test decoration, ideally based on any parameter that is passed to the test function rather than just `device` / `dtype`.

Case in point: #78735 is adding support for running tests across train / eval modes for `ModuleInfo`. Some tests are expected to fail for eval mode only, and there's no granular way to indicate this.

**Comment by zou3519** [comment]
This seems difficult to add, what would the API look like? Would the user provide a lambda that, on accepting the parameters, return a decorator?

**Comment by kshitij12345** [comment]
I think this will also help with the usage of `@parametrize` once they become frequent.

https://github.com/pytorch/pytorch/blob/a299a2fa262ba26a2aa3519d2b6d27a33e82f580/test/test_ops.py#L515-L518

For the test above, if an operator fails only for one of the executor, we have no choice but to add skip. (With expectedFailure, as the other executor passes, you get unexpected success)

**Comment by jbschlosser** [comment]
@zou3519 I've gotten a PoC working locally that adds the support by enhancing the `active_if` arg that can be used with `DecorateInfo`. Usage is similar to what you've described - a lambda that accepts a dict with the params being passed to the test -> returning a bool for whether the `DecorateInfo` is active.

For the example from @kshitij12345, the relevant `OpInfo` entry would contain something like this:
```python
DecorateInfo(
    unittest.expectedFailure, "TestOps", "test_python_ref_executor",
    device_type='cuda', active_if=lambda params: params['executor'] == 'nvfuser'
),
```

In the current code, the test decoration / skipping mechanisms are implemented somewhat redundantly in `OpInfo` and `ModuleInfo`, although they both use `DecorateInfo`. To support decoration based on any test param, I made this mechanism global instead so it's aware of the final state of the test function after composing decorators, etc. In my PoC, it's handled during test instantiation by `instantiate_parametrized_tests()` / `instantiate_device_type_tests()`.

A nice side effect of making the mechanism global is that it opens up the possibility for usage outside of `@ops` / `@modules` as well. It allows for things like this:

```python
# Example: Expect failure if x == 3 and we're in eval mode.
@applyIf(unittest.expectedFailure,
         lambda params: params['x'] == 3 and not params['training'])
@parametrize("x", range(5))
@parametrize("training", [False, True])
def test_foo(self, x, training):
    ...
```

Open to bikeshedding on how the API could look :)

**Comment by Lezcano** [comment]
Note that this would be very useful for regular OpInfos. 

There are some operations that have valid outputs, but that are not differentiable on those outputs. The most recent example I've bumped into is `logdet` which returns `NaN` if the input is real and has negative determinant. We could potentially skip gradchecks for these inputs, while still test them in all the other tests without having ot write and maintain a new OpInfo for them.
